src/ca/dbdToPv.cpp

```diff
@@ -0,0 +1,938 @@
+            case DBR_STS_DOUBLE:
+            {
+                const dbr_sts_double *data = static_cast<const dbr_sts_double *>(args.dbr);
+                status = data->status;
+                severity = data->severity;
+                break;
+            }
+            default:
+                throw  std::runtime_error("DbdToPv::getDone logic error");
+        }       
+    }
+    if(alarmRequested) {
+         bool statusChanged = false;
+         bool severityChanged = false;
+         PVStructurePtr pvAlarm(pvStructure->getSubField<PVStructure>("alarm"));
+         PVIntPtr pvSeverity(pvAlarm->getSubField<PVInt>("severity"));
+         if(caAlarm.severity!=severity) {
+             caAlarm.severity = severity;
+             pvSeverity->put(severity);
+             severityChanged = true;
+         }
+         PVStringPtr pvMessage(pvAlarm->getSubField<PVString>("message"));
+         PVIntPtr pvStatus(pvAlarm->getSubField<PVInt>("status"));
+         if(caAlarm.status!=status) {
+             caAlarm.status = status;
+             pvStatus->put(status);
+             string message("UNKNOWN STATUS");
+             if(status<=ALARM_NSTATUS) message = string(epicsAlarmConditionStrings[status]);
+             pvMessage->put(message);
+             statusChanged = true;
+         }
+         if(statusChanged&&severityChanged) {
+             bitSet->set(pvAlarm->getFieldOffset());
+         } else if(severityChanged) {
+             bitSet->set(pvSeverity->getFieldOffset());
+         } else if(statusChanged) {
+             bitSet->set(pvStatus->getFieldOffset());
+             bitSet->set(pvMessage->getFieldOffset());
+         }
+    }
+    if(timeStampRequested) {
+        PVStructurePtr pvTimeStamp(pvStructure->getSubField<PVStructure>("timeStamp"));
+        if(caTimeStamp.secPastEpoch!=stamp.secPastEpoch) {
+            caTimeStamp.secPastEpoch = stamp.secPastEpoch;
+            PVLongPtr pvSeconds(pvTimeStamp->getSubField<PVLong>("secondsPastEpoch"));
+            pvSeconds->put(stamp.secPastEpoch+posixEpochAtEpicsEpoch);
+            bitSet->set(pvSeconds->getFieldOffset());
+        }
+        if(caTimeStamp.nsec!=stamp.nsec) {
+            caTimeStamp.secPastEpoch = stamp.secPastEpoch;
+            PVIntPtr pvNano(pvTimeStamp->getSubField<PVInt>("nanoseconds"));
+            pvNano->put(stamp.nsec);
+            bitSet->set(pvNano->getFieldOffset());
+        }
+        
+    }
+    if(firstTime) {
+        firstTime = false;
+        bitSet->clear();
+        bitSet->set(0);
+    }
+    return Status::Ok;
+}
+
+Status DbdToPv::putToDBD(
+     CAChannelPtr const & caChannel,
+     PVStructurePtr const & pvStructure,
+     bool block)
+{
+    chid channelID = caChannel->getChannelID();
+    const void *pValue = NULL;
+    unsigned long count = 1;
+    dbr_enum_t   indexvalue(0);
+    dbr_char_t   bvalue(0);
+    dbr_short_t  svalue(0);
+    dbr_long_t   ivalue(0);
+    dbr_float_t  fvalue(0);
+    dbr_double_t dvalue(0);
```

Comment:
This would be a great situation for a C union.

https://github.com/epics-base/pva2pva/blob/1d0be31d3f96d33514b3e4cd795de48f430d8e26/pdbApp/pvif.h#L35-L46